src/storage/mvcc/reader/scanner/forward.rs

```diff
@@ -240,121 +302,129 @@ impl<S: Snapshot> ForwardScanner<S> {
         if needs_seek {
             // `user_key` must have reserved space here, so its clone has reserved space too. So no
             // reallocation happens in `append_ts`.
-            self.write_cursor
-                .seek(&user_key.clone().append_ts(ts), &mut self.statistics.write)?;
-            if !self.write_cursor.valid()? {
+            self.cursors.write.seek(
+                &user_key.clone().append_ts(self.cfg.ts),
+                &mut self.statistics.write,
+            )?;
+            if !self.cursors.write.valid()? {
                 // Key space ended.
-                return Ok(None);
+                return Ok(false);
             }
-            let current_key = self.write_cursor.key(&mut self.statistics.write);
+            let current_key = self.cursors.write.key(&mut self.statistics.write);
             if !Key::is_user_key_eq(current_key, user_key.as_encoded().as_slice()) {
                 // Meet another key.
-                *met_next_user_key = true;
-                return Ok(None);
+                return Ok(false);
             }
         }
+        Ok(true)
+    }
+}
 
-        // Now we must have reached the first key >= `${user_key}_${ts}`. However, we may
-        // meet `Lock` or `Rollback`. In this case, more versions needs to be looked up.
-        loop {
-            let write = WriteRef::parse(self.write_cursor.value(&mut self.statistics.write))
-                .map_err(Error::from)?;
-            self.statistics.write.processed += 1;
+/// `ForwardScanner` with this policy outputs the latest key value pairs.
+pub struct LatestKvPolicy;
+
+impl<S: Snapshot> ScanPolicy<S> for LatestKvPolicy {
+    type Output = (Key, Value);
+
+    fn handle_lock(
+        &mut self,
+        current_user_key: Key,
+        cfg: &mut ScannerConfig<S>,
+        cursors: &mut Cursors<S>,
+        statistics: &mut Statistics,
+    ) -> Result<HandleRes<Self::Output>> {
+        let result = match cfg.isolation_level {
+            IsolationLevel::Si => {
+                // Only needs to check lock in SI
+                let lock = {
+                    let lock_value = cursors.lock.value(&mut statistics.lock);
+                    Lock::parse(lock_value)?
+                };
+                lock.check_ts_conflict(&current_user_key, cfg.ts, &cfg.bypass_locks)
+                    .map(|_| ())
+            }
+            IsolationLevel::Rc => Ok(()),
+        };
+        cursors.lock.next(&mut statistics.lock);
+        // Even if there is a lock error, we still need to step the cursor for future
+        // calls.
+        if result.is_err() {
+            cursors.move_write_cursor_to_next_user_key(&current_user_key, statistics)?;
```

Comment:
when encountering conflict lock, now it moves write cursor to next user key no matter whether it has write for current user key. 
seems it doesn't follow the previous behavior.

Reply:
In the [original code](https://github.com/tikv/tikv/blob/26097303ded323637e3857fe12c871868e37af77/src/storage/mvcc/reader/scanner/forward.rs), if we meet a conflicting lock, `result`in L149 is set to an error. Then, L177 ~ L181 is skipped so `met_next_user_key` must be `false`. The write cursor is thus moved to the next user key.
I think I keep the behavior the same.

Reply:
oh, I misread the code. it will check current_user_key